pyrealm/demography/canopy_functions.py

```diff
@@ -0,0 +1,411 @@
+    crown_area: NDArray[np.float32],
+    f_g: NDArray[np.float32],
+    q_m: NDArray[np.float32],
+    z_max: NDArray[np.float32],
+    validate: bool = True,
+) -> NDArray[np.float32]:
+    """Calculate projected leaf area above a given height.
+
+    This function calculates the projected leaf area of a set of stems with given
+    properties at a set of vertical heights. This differs from crown area in allowing
+    for crown openness within the crown of an individual stem that results in the
+    displacement of leaf area further down into the canopy. The degree of openness is
+    controlled by the crown gap fraction property of each stem.
+
+    The stem properties are given in the arguments
+    ``stem_height``,``crown_area``,``f_g``,``q_m`` and ``z_max``, which must be
+    one-dimensional arrays ('row vectors') of equal length. The array of vertical
+    heights ``z`` accepts a range of input shapes (see
+    :meth:`~pyrealm.demography.canopy_functions.calculate_relative_canopy_radius_at_z`
+    ) and this function then also requires the expected relative stem radius (``q_z``)
+    calculated from those heights.
+
+    Args:
+        z: Vertical heights on the z axis.
+        q_z: Relative crown radius at heights in z.
+        crown_area: Crown area for a stem
+        stem_height: Total height of a stem
+        f_g: Within crown gap fraction for each stem.
+        q_m: Canopy shape parameter ``q_m``` for each stem
+        z_max: Height of maximum canopy radius for each stem
+        validate: Boolean flag to suppress argument validation.
+    """
+
+    # TODO - could this merge with the stem crown area function? A lot of overlap, so
+    # could be much more efficient to return both from one function.
+
+    if validate:
+        _validate_z_args(z, crown_area, stem_height, f_g, q_m, z_max)
+        _validate_q_z(z, q_z, crown_area)
+
+    # Calculate Ac terms
+    A_c_terms = crown_area * (q_z / q_m) ** 2
+
+    # Set Acp either side of z_max
+    A_cp = np.where(
+        z <= z_max,
+        crown_area - A_c_terms * f_g,
+        A_c_terms * (1 - f_g),
+    )
+    # Set Ap = 0 where z > H
+    A_cp = np.where(z > stem_height, 0, A_cp)
+
+    return A_cp
+
+
+# def calculate_total_canopy_A_cp(z: float, f_g: float, community: Community) -> float:
+#     """Calculate total leaf area at a given height.
+
+#     :param f_g:
+#     :param community:
+#     :param z: Height above ground.
+#     :return: Total leaf area in the canopy at a given height.
+#     """
+#     A_cp_for_individuals = calculate_projected_leaf_area_for_individuals(
+#         z, f_g, community
+#     )
+
+#     A_cp_for_cohorts = A_cp_for_individuals * community.cohort_number_of_individuals
+
+#     return A_cp_for_cohorts.sum()
+
+
+# def calculate_gpp(cell_ppfd: NDArray, lue: NDArray) -> float:
+#     """Estimate the gross primary productivity.
+
+#     Not sure where to place this - need an array of LUE that matches to the
+
+#     """
+
+#     return 100
```

Comment:
Can these be removed?

Reply:
They're placeholders for the next steps in this milestone, so easier for me to leave them there for now.